### PR TITLE
Fix hazard free

### PR DESCRIPTION
**Pull request by kumpera**
Replace usage of mono_thread_hazardous_free_or_queue and remove it.
    
    mono_thread_hazardous_free_or_queue has a fundamentally broken design.
    
    It allows arbitrary free code to run in the context of its caller.
    
    The sync/async split on where it's called is not enough to know whether
    we can run free'ing code or not.
    
    Since in sync context we called free functions that could take locks,
    it happened that those locks conflicted with the ones already taken.
    
    In particular, mono_jit_info_table_free did take a domain lock and
    quite a few of the sync context calls would hold locks that must
    not be held when taking a domain lock.
    
    This, is practice, means that we'd need to partition the free calls
    into 3 groups: async context, reentrant context (no runtime locks held)
    and sync context (maybe some locks held).
    
    There was no case where reentrant context would be usable meaning,
    in practice, that all calls happens in what effectively is async context
    where free functions can't be called.
    
    The new design is a lot more straightforward:
    
    - No implicit free queue pumping
    - A pair of free functions with well defined behavior.
    	mono_thread_hazardous_try_free - to be used where the caller expects the free function to be called
    	mono_thread_hazardous_queue_free - to be used where the caller don't expect the free function to be called [1]
    - Explicit pumping on places that known to be ok
    	Thread detach
    	Finalizer thread
    
    [1] This might sound like a weird condition, but a lot of lock-free code
    have compensation logic that trigger free'ing during lookups.

### CODE REVIEW

File: mono/utils/mono-linked-list-set.c [hunk 1]
```diff
@@ -126,7 +126,7 @@ mono_lls_find (MonoLinkedListSet *list, MonoThreadHazardPointers *hp, uintptr_t
 				mono_memory_write_barrier ();
 				mono_hazard_pointer_clear (hp, 1);
 				if (list->free_node_func)
-					mono_thread_hazardous_free_or_queue (cur, list->free_node_func, list->locking, context);
+					mono_thread_hazardous_queue_free (cur, list->free_node_func);
 			} else
```
Comment (vargaz) [inline mono/utils/mono-linked-list-set.c]: Why can't we free here ?

**Comment by vargaz** [comment]
LGTM. Without signalling the finalizer thread, won't these queues reach arbitrary length ?

File: mono/utils/mono-linked-list-set.c (hunk 1 above)
Comment (kumpera) [inline mono/utils/mono-linked-list-set.c]: This is a conservative assumption about free not being good outside of _remove.

This only happens under extreme load so it's ok to have it.

**Comment by kumpera** [comment]
Yes, the free queue can reach an arbitrary size, but here's the catch, all the callee before were not good spots for it.

What's I'll add is a way for the finalizer thread to be signaled if the queue goes above a certain threshold.

**Comment by vargaz** [comment]
LGTM.

**Comment by lewurm** [comment]
@monojenkins merge

**Pull request merged**